etc/init.d/zfs-load-key.in

```diff
@@ -0,0 +1,162 @@
+#!@DEFAULT_INIT_SHELL@
+#
+# zfs-load-key  This script will load/unload the zfs filesystems keys.
+#
+# chkconfig:    2345 06 99
+# description:  This script will load or unload the zfs filesystems keys during
+#               system boot/shutdown. Only filesystems with key path set
+#               in keylocation property. See the zfs(8) man page for details.
+# probe: true
+#
+### BEGIN INIT INFO
+# Provides:          zfs-load-key
+# Required-Start:    $local_fs zfs-import
+# Required-Stop:     $local_fs zfs-import
+# Default-Start:     2 3 4 5
+# Default-Stop:      0 1 6
+# X-Start-Before:    zfs-mount
+# X-Stop-After:      zfs-zed
+# Short-Description: Load ZFS keys for filesystems and volumes
+# Description: Run the `zfs load-key` or `zfs unload-key` commands.
+### END INIT INFO
+#
+# Released under the 2-clause BSD license.
+#
+# The original script that acted as a template for this script came from
+# the Debian GNU/Linux kFreeBSD ZFS packages (which did not include a
+# licensing stansa) in the commit dated Mar 24, 2011:
+#   https://github.com/zfsonlinux/pkg-zfs/commit/80a3ae582b59c0250d7912ba794dca9e669e605a
+
+# Source the common init script
+. @sysconfdir@/zfs/zfs-functions
+
+# ----------------------------------------------------
+
+do_depend()
+{
+	# bootmisc will log to /var which may be a different zfs than root.
+	before bootmisc logger zfs-mount
+
+	after zfs-import sysfs
+	keyword -lxc -openvz -prefix -vserver
+}
+
+# Load keys for all datasets/filesystems
+do_load_keys()
+{
+	local encryptionroot
+
+	zfs_log_begin_msg "Load ZFS filesystem(s) keys"
+
+	for encryptionroot in $("$ZFS" get -H -o value encryptionroot | uniq)
```

Comment:
This is broken: the result of the process substitution can be, e.g.
```
pool/data set
pool/data set/very cool
```
which will tokenise as `pool/data`, `set`, `pool/data`, `set/very`, `cool`. Set IFS to newline before the for.

Also: what about if encroots are `p/a`, `p/a/b`, `p/a`?
Also: you seem to be using another two zfs processes per iteration; why not `zfs list -Ho name,encryptionroot,keystatus,keylocation | while IFS="[TAB]" read -r name encroot kstat kloc; do if [ "$encroot" = "-" ] || [ "$name" != "$encroot"] || [ "$kloc" != "unavailable" ]; then continue; fi; ....; done`? That should take care of inter-dataset ordering, too, and reverses well.